### PR TITLE
Add arg to skip `with()` in `object_usage_linter()`

### LINKED ISSUE #1458: option to disable the `object_usage_linter` inside `with` expressions
would it be possible to add an option to the `object_usage_linter` so it does not check the inside of `with` expressions?

running `lintr::lint("this_file.R")`

```r
#this_file.R
test_fun <- function(df) {
  with(df, first_var + second_var)
}
```

throw the following warnings: 

```
this_file.R:2:12: warning: [object_usag e_linter] no visible binding for global variable ‘first_var’ with(df, first_var + second_var)
this_file.R:2:24: warning: [object_usag e_linter] no visible binding for global variable ‘second_var’ with(df, first_var + second_var)
```

but strangely no such warning appear with this code:

```r
df <- readRDS("df_path.rds")
with(df, first_var + second_var)
```

Thanks

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
Closes #1458

# Example

``` r
library(lintr)

f <- tempfile()
writeLines(
  "test_fun <- function(df) {
      with(df, first_var + second_var)
     }",
  f
)

lint(f, object_usage_linter(skip_with = TRUE))

lint(f, object_usage_linter(skip_with = FALSE))
#> /private/var/folders/xr/v_vddzvs33q5wg7jv9mr__4h0000gn/T/Rtmp1OPl2d/file1557f756350e9:2:16: warning: [object_usage_linter] no visible binding for global variable 'first_var'
#>       with(df, first_var + second_var)
#>                ^~~~~~~~~
#> /private/var/folders/xr/v_vddzvs33q5wg7jv9mr__4h0000gn/T/Rtmp1OPl2d/file1557f756350e9:2:28: warning: [object_usage_linter] no visible binding for global variable 'second_var'
#>       with(df, first_var + second_var)
#>                            ^~~~~~~~~~
```

<sup>Created on 2022-09-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

P.S. A few `.Rd` changes are remnants of #1539, I think.

### CODE REVIEW

**Comment by AshesITR** [comment]
Isn't R CMD checks default `skipWith = TRUE`? Are there valid usages of `with()` that don't produce false positives?

**Comment by IndrajeetPatil** [comment]
@AshesITR You are right; I have changed the default for `skip_with` to `TRUE`, since setting is to `FALSE` is rarely going to _not_ produce a false positive.

I also made some changes to the docs to close https://github.com/r-lib/lintr/issues/941. The rest of the todos in that issue already seem to have been handled.

File: R/object_usage_linter.R [hunk 1]
```diff
@@ -5,11 +5,12 @@
 #'
 #' @param interpret_glue If `TRUE`, interpret [glue::glue()] calls to avoid false positives caused by local variables
 #' which are only used in a glue expression.
+#' @param skip_with A logical. If `TRUE` (default), code in `with()` expressions will be skipped.
```
Comment (MichaelChirico) [inline R/object_usage_linter.R:8]: lets mention this is passed along to the `skipWith` argument to remove any ambiguity

File: R/object_usage_linter.R [hunk 2]
```diff
@@ -197,7 +199,7 @@ get_assignment_symbols <- function(xml) {
 }
 
 parse_check_usage <- function(expression, known_used_symbols = character(), declared_globals = character(),
-                              start_line = 1L) {
+                              start_line = 1L, skip_with = FALSE) {
```
Comment (MichaelChirico) [inline R/object_usage_linter.R:202]: any reason not to change the default here too?

File: tests/testthat/test-object_usage_linter.R [hunk 3]
```diff
@@ -480,3 +480,15 @@ test_that("unknown infix operators give good lint metadata", {
...
+
+test_that("respects `skip_with` argument for `with()` expressions", {
+  f <- tempfile()
+  writeLines(
+    "test_fun <- function(df) {
+      with(df, first_var + second_var)
+     }",
+    f
+  )
+  expect_true(length(lint(f, object_usage_linter(skip_with = TRUE))) == 0L)
```
Comment (MichaelChirico) [inline tests/testthat/test-object_usage_linter.R:492]: use `expect_length()` here

File: tests/testthat/test-object_usage_linter.R [hunk 4]
```diff
@@ -480,3 +480,15 @@ test_that("unknown infix operators give good lint metadata", {
     object_usage_linter()
   )
 })
+
+test_that("respects `skip_with` argument for `with()` expressions", {
+  f <- tempfile()
```
Comment (MichaelChirico) [inline tests/testthat/test-object_usage_linter.R:485]: this `f` is not cleaned up afterwards.

better to use `withr::local_tempfile(lines = c("test_fun...", ...))`

File: R/object_usage_linter.R (hunk 2 above)
Comment (IndrajeetPatil) [reply R/object_usage_linter.R]: No reason; just a leftover from the original approach I had taken in the PR. 

Changed to `skip_with  = TRUE`.

File: tests/testthat/test-object_usage_linter.R (hunk 4 above)
Comment (IndrajeetPatil) [reply tests/testthat/test-object_usage_linter.R]: Good point. Changed to use `{withr}`.

File: tests/testthat/test-object_usage_linter.R (hunk 3 above)
Comment (IndrajeetPatil) [reply tests/testthat/test-object_usage_linter.R]: Done.

Should we be including the `expect_*()` linters in lint changed files GHA?

Comment (MichaelChirico) [reply tests/testthat/test-object_usage_linter.R]: I was wondering the same... works for me

File: tests/testthat/test-object_usage_linter.R [hunk 5]
```diff
@@ -482,13 +482,13 @@ test_that("unknown infix operators give good lint metadata", {
...
-    "test_fun <- function(df) {
-      with(df, first_var + second_var)
-     }",
-    f
-  )
-  expect_true(length(lint(f, object_usage_linter(skip_with = TRUE))) == 0L)
-  expect_true(length(lint(f, object_usage_linter(skip_with = FALSE))) == 2L)
+  f <- withr::local_tempfile(
+    lines =
+      "test_fun <- function(df) {
```
Comment (MichaelChirico) [inline tests/testthat/test-object_usage_linter.R:487]: better as a character vector of lines IMO

File: tests/testthat/test-object_usage_linter.R (hunk 3 above)
Comment (IndrajeetPatil) [reply tests/testthat/test-object_usage_linter.R]: Done: https://github.com/r-lib/lintr/pull/1557

File: tests/testthat/test-object_usage_linter.R (hunk 5 above)
Comment (IndrajeetPatil) [reply tests/testthat/test-object_usage_linter.R]: Done.

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**